xmake/rules/c++/modules/modules_support/gcc.lua

```diff
@@ -149,77 +149,91 @@ function generate_dependencies(target, sourcebatch, opt)
     end
 end
 
--- generate target header units
-function generate_headerunits(target, batchcmds, headerunits, opt)
+-- generate target stl header units
+function generate_stl_headerunits(target, batchcmds, headerunits, opt)
     local compinst = target:compiler("cxx")
-    local cachedir = common.modules_cachedir(target)
-    local stlcachedir = common.stlmodules_cachedir(target)
     local mapper_file = _get_module_mapper()
 
+    -- get cachedirs
+    local stlcachedir = common.stlmodules_cachedir(target)
+
     -- build headerunits
-    local objectfiles = {}
     local projectdir = os.projectdir()
-    for _, headerunit in ipairs(headerunits) do
-        if not headerunit.stl then
-            local file = path.relative(headerunit.path, target:scriptdir())
-            local objectfile = target:objectfile(file)
+    for i, headerunit in ipairs(headerunits) do
+        local bmifile = path.join(stlcachedir, headerunit.name .. get_bmi_extension())
+        if _add_module_to_mapper(mapper_file, headerunit.path, path.absolute(bmifile, projectdir)) then
+            local args = { "-c", "-x", "c++-system-header", headerunit.name }
 
-            local outdir
-            if headerunit.type == ":quote" then
-                outdir = path.join(cachedir, path.directory(path.relative(headerunit.path, projectdir)))
-            else
-                outdir = path.join(cachedir, path.directory(headerunit.path))
-            end
-            if not os.isdir(outdir) then
-                os.mkdir(outdir)
-            end
+            batchcmds:show_progress(opt.progress, "${color.build.object}generating.cxx.headerunit.bmi %s", headerunit.name)
+            batchcmds:vrunv(compinst:program(), table.join(compinst:compflags({target = target}), args))
 
-            local bmifilename = path.basename(objectfile) .. get_bmi_extension()
-            local bmifile = (outdir and path.join(outdir, bmifilename) or bmifilename)
-            if not os.isdir(path.directory(objectfile)) then
-                os.mkdir(path.directory(objectfile))
-            end
+            batchcmds:set_depmtime(os.mtime(bmifile))
+            batchcmds:set_depcache(target:dependfile(bmifile))
```

Comment:
We should not set it multiple times in the loop

Reply:
so i can't use set_depmtime/set_depcache because there is a lot of bmifile ?

Reply:
yes, they should be set once at the end

```lua
    before_buildcmd_files(function(target, batchcmds, sourcebatch, opt)
         for loop
               batchcmds:vrunv()
               batchcmds:vrunv()
         end
         batchcmds:set_depmtime(math.max(all_bmi_files_mtimes))
         batchcmds:set_depcache(target:dependfile("all_bmi_files ..."))
    end)
```

Unless we use `before_buildcmd_file` and build the bmi file separately for each sourcefile, then we set mtime/depcache for each bmifile, but there will be parallel compilation problems.

This is because each file job is executed in parallel.

Reply:
So, while before_buildcmd_files are usually used for project generators and simple tasks, before_build_files can handle tasks in a more flexible way.